pkg/standalone-utils/test/BaseRelayer.test.ts

```diff
@@ -0,0 +1,168 @@
+          relayer: relayer.address,
+          sender: signer.address,
+          approved: false,
+        });
+      });
+    });
+
+    context('when relayer is not allowed to set approval', () => {
+      it('reverts', async () => {
+        await expect(relayer.connect(signer).setRelayerApproval(true, approvalAuthorisation)).to.be.revertedWith(
+          'SENDER_NOT_ALLOWED'
+        );
+      });
+    });
+  });
+
+  describe('multicall', () => {
+    function setRelayerApprovalTx(allowed: boolean, authorization: string) {
+      return relayer.interface.encodeFunctionData('setRelayerApproval', [allowed, authorization]);
+    }
+
+    context('when sending ETH', () => {
+      it('refunds the unused ETH', async () => {
+        // Pass in 100 wei which will not be used
+        const value = 100;
+        const userBalanceBefore = await ethers.provider.getBalance(signer.address);
+
+        const tx = await relayer.connect(signer).multicall([], { value });
+        const receipt = await tx.wait();
+
+        const txCost = tx.gasPrice.mul(receipt.gasUsed);
+        const expectedBalanceAfter = userBalanceBefore.sub(txCost);
+        const userBalanceAfter = await ethers.provider.getBalance(signer.address);
+
+        expect(userBalanceAfter).to.be.eq(expectedBalanceAfter);
+        expect(await ethers.provider.getBalance(vault.address)).to.be.eq(0);
+        expect(await ethers.provider.getBalance(relayer.address)).to.be.eq(0);
+      });
+    });
+
+    context('when passed a call which will revert', () => {
+      it('passes up the correct revert string', async () => {
+        // Call should fail due to relayer not been approved by the Authorizer
+        await expect(
+          relayer.connect(signer).multicall([setRelayerApprovalTx(true, approvalAuthorisation)])
+        ).to.be.revertedWith('SENDER_NOT_ALLOWED');
+
+        const authorizer = vault.authorizer as Contract;
+        const setRelayerApproval = actionId(vault.instance, 'setRelayerApproval');
+        await authorizer.connect(admin).grantRole(setRelayerApproval, relayer.address);
+
+        // Call should fail due to bad authorization from user
+        await expect(
+          relayer.connect(signer).multicall([setRelayerApprovalTx(true, EMPTY_AUTHORISATION)])
+        ).to.be.revertedWith('USER_DOESNT_ALLOW_RELAYER');
+      });
+    });
+
+    context('when relayer is allowed to set approval', () => {
+      sharedBeforeEach('authorise relayer', async () => {
+        const authorizer = vault.authorizer as Contract;
+        const setRelayerApproval = actionId(vault.instance, 'setRelayerApproval');
+        await authorizer.connect(admin).grantRole(setRelayerApproval, relayer.address);
+      });
+
+      context('when not approved by sender', () => {
+        context('when the first call gives permanent approval', () => {
+          it("doesn't require signatures on further calls", async () => {
```

Comment:
The line between Vault and BaseRelayer tests is a bit blurred here, but these tests are simple enough that I think it can be useful to keep them.

Reply:
Agreed, although considering this is a core feature of this relayer it's nice to have direct confirmation that this works.